File: src/main/java/se/mejsla/java/account/AccountSummary.java
package se.mejsla.java.account;

/**
 * Contains the name and balance of one account.
 */
public class AccountSummary {

    private String name;

    private int balance;

    public AccountSummary(String name, int balance) {
        this.name = name;
        this.balance = balance;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public int getBalance() {
        return balance;
    }

    public void setBalance(int balance) {
        this.balance = balance;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        AccountSummary that = (AccountSummary) o;

        if (balance != that.balance) return false;
        if (name != null ? !name.equals(that.name) : that.name != null) return false;

        return true;
    }

    @Override
    public int hashCode() {
        int result = name != null ? name.hashCode() : 0;
        result = 31 * result + balance;
        return result;
    }

    @Override
    public String toString() {
        return getClass().getSimpleName() + "(" + name + "," + balance + ")";
    }
}

File: src/main/scala/se/mejsla/scala/account/AccountSummary.scala
package se.mejsla.scala

/*
 * One version of a more idiomatic counterpart of the se.mejsla.java.account classes. Note that they could be put in separate files just
 * like the Java versions but since it is so little code they are in the same file which is legal in Scala
 */
package object account {


  /**
   * Contains the name and balance of one account.
   */
  case class AccountSummary(name: String, balance: Int)

  /**
   * Contains customer name and all the account summaries for one customer
   */
  case class CustomerSummary(name: String, accountSummaries: Seq[AccountSummary])

  /**
   * @return The balance of the given customer and account or None if the account or user does not exist
   */
  def balanceFor(customerSummaries: => Seq[CustomerSummary])(customerName: String, accountName: String): Option[Int] =
    customerSummaries.find(_.name == customerName).flatMap { customerSummary =>
      customerSummary.accountSummaries.find(_.name == accountName).map(_.balance)
    }

}




